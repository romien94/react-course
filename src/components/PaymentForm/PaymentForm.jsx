import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { saveProfile, loadProfile } from "../../modules/actions";

import FormLabel from "../common/FormLabel";
import Input from "../common/Input";
import Button from "../common/Button";

class PaymentForm extends React.Component {
  static propTypes = {
    saveProfile: PropTypes.func,
    loadProfile: PropTypes.func,
  };

  state = {
    number: "",
    date: "",
    name: "",
    cvc: "",
  };

  formRef = React.createRef();

  componentDidMount() {
    this.props.loadProfile(this.props.token);
    const { number, date, name, cvc } = this.props;
    this.setState({
      number,
      date,
      name,
      cvc,
    });
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props !== prevProps) {
      const { number, date, name, cvc } = this.props;
      this.setState({
        number,
        date,
        name,
        cvc,
      });
    }
  }

  render() {
    return (
      <form
        ref={this.formRef}
        onSubmit={async (e) => {
          e.preventDefault();
          const { number, date, name, cvc } = this.state;
          const { token } = this.props;
          this.props.saveProfile(number, date, name, cvc, token);
        }}
        className="app-profile__form app-form"
      >
        <div className="app-form__wrapper">
          <div className="app-form__cols">
            <div className="app-form__col">
              <div className="app-form__row">
                <FormLabel>
                  <span className="app-form__fieldname">Номер карты:</span>
                  <Input
                    onChange={(e) => {
                      const value = e.target.value;
                      this.setState({ number: value });
                    }}
                    type="text"
                    name="number"
                    data-testid="number"
                    value={this.state.number}
                    className="app-form__input"
                  />
                </FormLabel>
              </div>
              <div className="app-form__row">
                <FormLabel>
                  <span className="app-form__fieldname">Срок действия:</span>
                  <Input
                    onChange={(e) => {
                      const value = e.target.value;
                      this.setState({ date: value });
                    }}
                    type="text"
                    name="date"
                    data-testid="date"
                    className="app-form__input"
                    value={this.state.date}
                  />
                </FormLabel>
              </div>
            </div>
            <div className="app-form__col">
              <div className="app-form__row">
                <FormLabel>
                  <span className="app-form__fieldname">Имя владельца:</span>
                  <Input
                    onChange={(e) => {
                      const value = e.target.value;
                      this.setState({ name: value });
                    }}
                    type="text"
                    name="name"
                    data-testid="name"
                    className="app-form__input"
                    value={this.state.name}
                  />
                </FormLabel>
              </div>
              <div className="app-form__row">
                <FormLabel>
                  <span className="app-form__fieldname">CVC:</span>
                  <Input
                    onChange={(e) => {
                      const value = e.target.value;
                      this.setState({ cvc: value });
                    }}
                    name="cvc"
                    data-testid="cvc"
                    type="password"
                    value={this.state.cvc}
                  />
                </FormLabel>
              </div>
            </div>
          </div>
          <div className="app-form__controls">
            <div className="app-form__row">
              <Button
                data-testid="save"
                type="submit"
                className="button app-form__button"
              >
                Сохранить
              </Button>
            </div>
          </div>
        </div>
      </form>
    );
  }
}

export default connect(
  (state) => ({ ...state.profile, token: state.auth.token }),
  { saveProfile, loadProfile }
)(PaymentForm);
