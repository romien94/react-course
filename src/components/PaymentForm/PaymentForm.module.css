.paymentForm {
    width: 100%;
    height: 100%;
}

.formCols {
    display: flex;
    margin-bottom: 45px;
}

.formCol {
    position: relative;
    display: flex;
    flex-direction: column;
    flex: 1;
    margin-right: 20px;
    max-width: 380px;
    /* border: 1px solid gray; */
    box-shadow: 0px 3px 3px -2px rgba(0, 0, 0, 0.2), 0px 3px 4px 0px rgba(0, 0, 0, 0.14),
        0px 1px 8px 0px rgba(0, 0, 0, 0.12);
    border-radius: 10px;
    padding: 30px;
    min-height: 231px;
    justify-content: space-between;
}

.formCol:first-child:after {
    position: absolute;
    top: 0;
    right: 0;
    content: "";
    display: block;
    width: 62px;
    height: 33px;
    background-image: url("../../assets/images/maste-card.png");
    background-repeat: no-repeat;
}

.formCol:last-child {
    margin-right: 0;
}

.formRow {
    display: flex;
    /* margin-bottom: 35px; */
}

.formRow:last-child {
    margin-bottom: 0;
}

.formLink {
    color: #0074ea;
    text-decoration: none;
}

.formLabel {
    display: flex;
    flex-direction: column;
    flex: 1;
}

.formInput {
    border: none;
    background-color: transparent;
    border-bottom: 1px solid gray;
    padding: 7px 0;
    font-size: 24px;
}


.formName {
    text-align: left;
}

.formLabel:hover .formInput {
    border-bottom-color: #0074ea;    
}

.formInput:active,
.formInput:focus {
    outline: none;
    border-bottom-color: #ffc700;
}



.formControls {
    display: flex;
    justify-content: center;
}
