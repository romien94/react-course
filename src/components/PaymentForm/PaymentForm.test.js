import React from "react";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import PaymentForm from "./PaymentForm";

describe("Login works fine", () => {
  const { getByTestId } = render(<PaymentForm />);
  it("Snapshot matches", () => {
    const mockPaymentForm = renderer.create(<PaymentForm />).toJSON();
    expect(mockPaymentForm).toMatchSnapshot();
  });

  describe("Contains all the necessary inputs with basic setup", () => {
    expect(getByTestId("number")).toBeTruthy();
    expect(getByTestId("date")).toBeTruthy();
    expect(getByTestId("name")).toBeTruthy();
    expect(getByTestId("cvc")).toBeTruthy();
    expect(getByTestId("save")).toBeTruthy();
    expect(getByTestId("save")).toHaveAttribute('type', 'submit');
  });
});
