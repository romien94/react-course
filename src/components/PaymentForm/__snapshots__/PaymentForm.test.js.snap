// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login works fine Snapshot matches 1`] = `
<form
  className="app-profile__form app-form"
>
  <div
    className="app-form__wrapper"
  >
    <div
      className="app-form__cols"
    >
      <div
        className="app-form__col"
      >
        <div
          className="app-form__row"
        >
          <label
            className="MuiFormLabel-root"
          >
            <span
              className="app-form__fieldname"
            >
              Номер карты:
            </span>
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline app-form__input"
              data-testid="number"
              onClick={[Function]}
            >
              <input
                className="MuiInputBase-input MuiInput-input"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="app-form__row"
        >
          <label
            className="MuiFormLabel-root"
          >
            <span
              className="app-form__fieldname"
            >
              Срок действия:
            </span>
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline app-form__input"
              data-testid="date"
              onClick={[Function]}
            >
              <input
                className="MuiInputBase-input MuiInput-input"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
              />
            </div>
          </label>
        </div>
      </div>
      <div
        className="app-form__col"
      >
        <div
          className="app-form__row"
        >
          <label
            className="MuiFormLabel-root"
          >
            <span
              className="app-form__fieldname"
            >
              Имя владельца:
            </span>
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline app-form__input"
              data-testid="name"
              onClick={[Function]}
            >
              <input
                className="MuiInputBase-input MuiInput-input"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="app-form__row"
        >
          <label
            className="MuiFormLabel-root"
          >
            <span
              className="app-form__fieldname"
            >
              CVC:
            </span>
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline"
              data-testid="cvc"
              onClick={[Function]}
            >
              <input
                className="MuiInputBase-input MuiInput-input"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="password"
              />
            </div>
          </label>
        </div>
      </div>
    </div>
    <div
      className="app-form__controls"
    >
      <div
        className="app-form__row"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained button app-form__button"
          data-testid="save"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
            Сохранить
          </span>
        </button>
      </div>
    </div>
  </div>
</form>
`;
